test/contrib/test_tfp.py

```diff
@@ -32,6 +32,7 @@ def test_api_consistent():
                 assert p in inspect.getfullargspec(tfp_dist.__init__)[0]
 
 
+@pytest.mark.filterwarnings("ignore:Importing distributions")
```

Comment:
This one is interesting. For completeness, it would be nice if you could add methods `tree_flatten`, `tree_unflatten` to `TFPDistribution`. Some sort of
```python
def tree_flatten(self):
    return jax.tree_util.tree_flatten(self.tfp_dist)

@classmethod
def tree_unflatten(cls, aux_data, params):
    fn = jax.tree_util.tree_unflatten(aux_data, params)
    return TFPDistribution[fn.__class__](**fn.parameters)
```

Reply:
While looking at this, I am a bit worried that we will keep registering the [pytree](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/distributions/distribution.py#L140) each time we have a `sample` statement. This might populate JAX pytree caches. I'll look at this in more detail later.

Reply:
I'm not sure how to check pytree caches, but I can confirm that the call to `register_pytree_node` is currently happening with every `numpyro.sample` statement.

Perhaps the wrapped distributions should be constructed in the module scope as before, and we can instantiate the correct one inside `numpyro.sample` using `getattr` on the model?

Reply:
It can be fixed by creating a cache `_TFD_CACHE = {}` in `contrib.tfp.distributions` and in the `__getitem__` method, return the cache if the tfd class is available there, otherwise caching the wrapped class at the end of `__getitem__`.

Reply:
I think we can just use `functools.lru_cache` for this with `maxsize=None`, what do you think? Have pushed a commit with this change. Tests are passing locally and I've verified that each distribution is only registered once.

Reply:
I tried adding the tree_flatten methods, running your snippet I get `TypeError: Value 'x' with type <class 'str'> is not a valid JAX type`. Same problem if I replace `tfd.Normal` with `numpyro.distributions.Normal`. 

Did you mean to trace inside the jitted function or jit _then_ trace perhaps?

Reply:
I guess the trace holds some non-device types such as `trace["type"] == "samples"`. You might just return `tr["fn"]` then :).

Reply:
> we can just use functools.lru_cache for this with maxsize=None

Nice solution! I didn't know this Python helper. Thanks!